Error/Timeout saving thing results in unusable thing

This is linked to https://github.com/openhab/openhab2-addons/issues/858

It seems that when saving the controller (bridge) in ZWave, if a frame is being sent (which can result in a 5 second timeout), then the updating of the thing times out (which is also 5 seconds), and this seems to leave the thing in an unusable state requiring it to be deleted.

When updating the thing, the following error occurs -:
```
2016-04-30 13:14:31.849 [DEBUG] [b.z.i.protocol.ZWaveController:527  ] - Message queued. Queue length = 5. Queue={}
2016-04-30 13:14:31.849 [DEBUG] [WaveController$ZWaveSendThread:1142 ] - Took message from queue for sending. Queue length = 4
2016-04-30 13:14:31.850 [DEBUG] [o.b.z.i.protocol.SerialMessage:255  ] - Assembled message buffer = 01 09 00 13 0E 02 63 04 25 0C A7 
2016-04-30 13:14:31.850 [DEBUG] [b.z.handler.ZWaveSerialHandler:305  ] - NODE 14: Sending REQUEST Message = 01 09 00 13 0E 02 63 04 25 0C A7 
2016-04-30 13:14:31.851 [DEBUG] [rialHandler$ZWaveReceiveThread:296  ] - Stopped ZWave thread: Receive
2016-04-30 13:14:34.919 [INFO ] [b.z.handler.ZWaveSerialHandler:130  ] - Stopped ZWave serial handler
2016-04-30 13:14:34.919 [DEBUG] [.o.b.z.d.ZWaveDiscoveryService:58   ] - Deactivating ZWave discovery service for zwave:serial_zstick:d01c0920
2016-04-30 13:14:34.922 [DEBUG] [b.z.handler.ZWaveSerialHandler:67   ] - Initializing ZWave serial controller.
2016-04-30 13:14:36.447 [DEBUG] [s.core.common.SafeMethodCaller:184  ] - Timeout of 5000ms exceeded, thread ESH-safeCall-4 (54) in state RUNNABLE is at org.openhab.binding.zwave.handler.ZWaveSerialHandler.initialize(ZWaveSerialHandler.java:69).
2016-04-30 13:14:36.449 [ERROR] [.c.thing.internal.ThingManager:525  ] - Exception occured while calling thing updated at ThingHandler 'org.openhab.binding.zwave.handler.ZWaveSerialHandler@4529295b: nulljava.util.concurrent.TimeoutException: null
	at java.util.concurrent.FutureTask.get(FutureTask.java:205)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.thingUpdated(ThingManager.java:515)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyTrackers(ThingRegistryImpl.java:218)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:141)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.updated(AbstractRegistry.java:102)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:57)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutUpdatedElement(AbstractProvider.java:82)

2016-04-30 13:14:36.449 [DEBUG] [.c.c.r.AbstractManagedProvider:135  ] - Updated element zwave:serial_zstick:d01c0920 in ManagedThingProvider.
2016-04-30 13:14:36.450 [INFO ] [rthome.event.ThingUpdatedEvent:43   ] - Thing 'zwave:serial_zstick:d01c0920' has been updated.
```

When the thing then `initialize()`s, the following code throws an NPE -:
```
        portId = (String) getConfig().get(CONFIGURATION_PORT);
```
presumable because the get returns null?

If OH is stopped and restarted, then the thing is never initialised -:
```
2016-04-30 13:19:20.563 [ERROR] [.c.thing.internal.ThingManager:598  ] - Exception occured while calling thing handler factory 'org.openhab.binding.zwave.internal.ZWaveHandlerFactory@57dc4cb0': java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridgejava.util.concurrent.ExecutionException: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.registerHandler(ThingManager.java:582)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.addThingHandlerFactory(ThingManager.java:902)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
Caused by: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
	at org.openhab.binding.zwave.internal.ZWaveHandlerFactory.createHandler(ZWaveHandlerFactory.java:50)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory.registerHandler(BaseThingHandlerFactory.java:113)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:586)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$6.call(ThingManager.java:1)
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	... 1 common frames omitted
```

In the zwave handler, this is here -:
```java
        // Handle controllers here
        if (thingTypeUID.equals(CONTROLLER_SERIAL)) {
            return new ZWaveSerialHandler((Bridge) thing);
        }
```

If I look at 'thing' it is type ```ThingImpl``` and i think it should be ```BridgeImpl```?

At this point, the only way out seems to be to delete the thing and start again...

The ZWave binding probably should exit quicker to avoid the timeout in the first place (I'll look at that), but if the timeout does occur, I think the system shouldn't die.

Fixing the thread termination in the ZWave receive didn't help -:

```
2016-04-30 17:16:39.077 [DEBUG] [b.z.handler.ZWaveSerialHandler:118  ] - Disposing of ZWave serial controller.
2016-04-30 17:16:39.776 [DEBUG] [rialHandler$ZWaveReceiveThread:297  ] - Stopped ZWave thread: Receive
2016-04-30 17:16:42.841 [INFO ] [b.z.handler.ZWaveSerialHandler:131  ] - Stopped ZWave serial handler
2016-04-30 17:16:42.842 [DEBUG] [o.b.z.h.ZWaveControllerHandler:174  ] - Disposing of ZWave Controller.
2016-04-30 17:16:42.842 [DEBUG] [WaveController$ZWaveSendThread:1271 ] - Stopped ZWave thread: Send
2016-04-30 17:16:42.843 [DEBUG] [b.z.i.protocol.ZWaveController:1115 ] - Stopped ZWave thread: Input
2016-04-30 17:16:42.845 [DEBUG] [.o.b.z.d.ZWaveDiscoveryService:58   ] - Deactivating ZWave discovery service for zwave:serial_zstick:154679680b2
2016-04-30 17:16:42.848 [DEBUG] [b.z.handler.ZWaveSerialHandler:67   ] - Initializing ZWave serial controller.
2016-04-30 17:16:42.863 [ERROR] [.c.thing.internal.ThingManager:525  ] - Exception occured while calling thing updated at ThingHandler 'org.openhab.binding.zwave.handler.ZWaveSerialHandler@30696784: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridgejava.util.concurrent.ExecutionException: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:206)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.thingUpdated(ThingManager.java:515)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyTrackers(ThingRegistryImpl.java:218)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:141)
	at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutUpdatedElement(ThingRegistryImpl.java:1)
	at org.eclipse.smarthome.core.common.registry.AbstractRegistry.updated(AbstractRegistry.java:102)
	at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:57)
Caused by: java.lang.ClassCastException: org.eclipse.smarthome.core.thing.internal.ThingImpl cannot be cast to org.eclipse.smarthome.core.thing.Bridge
	at org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler.getThing(BaseBridgeHandler.java:54)
	at org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler.getThing(BaseBridgeHandler.java:1)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.getConfig(BaseThingHandler.java:202)
	at org.openhab.binding.zwave.handler.ZWaveSerialHandler.initialize(ZWaveSerialHandler.java:69)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.thingUpdated(BaseThingHandler.java:161)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$5.call(ThingManager.java:519)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$5.call(ThingManager.java:1)
	at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)

2016-04-30 17:16:42.865 [DEBUG] [.c.c.r.AbstractManagedProvider:135  ] - Updated element zwave:serial_zstick:154679680b2 in ManagedThingProvider.
2016-04-30 17:16:42.866 [INFO ] [rthome.event.ThingUpdatedEvent:43   ] - Thing 'zwave:serial_zstick:154679680b2' has been updated.
```

I am seeing an exception in the IDE only when registering the discovery service in ZWave - I don't understand why as it looks ok, and ultimately it registers fine. I'm not sure if this is related - ie if the exception during initialisation is causing the thing not to be created properly...

--- a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/util/ThingHelper.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/util/ThingHelper.java
@@ -13,13 +13,17 @@ import java.util.Collections;
 import java.util.List;
 
 import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.core.thing.Bridge;
 import org.eclipse.smarthome.core.thing.Channel;
 import org.eclipse.smarthome.core.thing.Thing;
 import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.builder.BridgeBuilder;
+import org.eclipse.smarthome.core.thing.binding.builder.GenericThingBuilder;
 import org.eclipse.smarthome.core.thing.binding.builder.ThingBuilder;
 import org.eclipse.smarthome.core.thing.dto.ChannelDTO;
 import org.eclipse.smarthome.core.thing.dto.ChannelDTOMapper;
 import org.eclipse.smarthome.core.thing.dto.ThingDTO;
+import org.eclipse.smarthome.core.thing.internal.BridgeImpl;
 import org.eclipse.smarthome.core.thing.internal.ThingImpl;
 
 import com.google.common.base.Joiner;
@@ -102,7 +106,13 @@ public class ThingHelper {
      */
     public static Thing merge(Thing thing, ThingDTO updatedContents) {
 
-        ThingBuilder builder = ThingBuilder.create(thing.getThingTypeUID(), thing.getUID());
+        GenericThingBuilder<?> builder;
+
+        if (thing instanceof Bridge) {
+            builder = BridgeBuilder.create(thing.getThingTypeUID(), thing.getUID());
+        } else {
+            builder = ThingBuilder.create(thing.getThingTypeUID(), thing.getUID());
+        }
 
         // Update the label
         if (updatedContents.label != null) {
@@ -141,6 +151,17 @@ public class ThingHelper {
             builder.withChannels(thing.getChannels());
         }
 
-        return builder.build();
+        Thing mergedThing = builder.build();
+
+        // keep all child things in place on a merged bridge
+        if (mergedThing instanceof BridgeImpl && thing instanceof Bridge) {
+            Bridge bridge = (Bridge) thing;
+            BridgeImpl mergedBridge = (BridgeImpl) mergedThing;
+            for (Thing child : bridge.getThings()) {
+                mergedBridge.addThing(child);
+            }
+        }
+
+        return mergedThing;
     }
 }

